pontoon/tags/utils/base.py

```diff
@@ -0,0 +1,193 @@
+    version of the object.
+
+    The clone will be called with attributes listed in `self.clone_kwargs` as
+    kwargs. These can be overridden when creating the clone.
+
+    """
+    clone_kwargs = ()
+
+    def __init__(self, **kwargs):
+        for k in self.clone_kwargs:
+            setattr(self, k, kwargs.get(k))
+
+    def __call__(self, **kwargs):
+        clone_kwargs = dict(
+            (k, getattr(self, k))
+            for k
+            in self.clone_kwargs)
+        clone_kwargs.update(kwargs)
+        return self.__class__(**clone_kwargs)
+
+
+class FilteredData(Clonable):
+    """Base Tool for constructing and coalescing aggregating querysets
+
+    Descendants of this class will filter a queryset by mapping
+    self.filter_methods to methods on the class
+
+    The data is aggregated and then cached/coalesced to the
+    data property
+
+    It can be cloned to override filtering params
+    """
+    default_groupby = ()
+    _default_annotations = ()
+
+    def __iter__(self):
+        return iter(self.data)
+
+    def __len__(self):
+        return len(self.data)
+
+    @property
+    def data_manager(self):
+        raise NotImplementedError()
+
+    @property
+    def default_annotations(self):
+        return OrderedDict(self._default_annotations)
+
+    @property
+    def filtered_data(self):
+        data = self.data_manager.all()
+        for tag_filter in self.filters:
+            data = tag_filter(data)
+        return data
+
+    @property
+    def filters(self):
+        return [
+            getattr(self, "filter_%s" % f)
+            for f
+            in self.filter_methods]
+
+    @cached_property
+    def data(self):
+        return self.coalesce(self.get_data())
+
+    def coalesce(self, data):
+        return data
+
+    def get_annotations(self):
+        anno = self.default_annotations.copy()
+        anno.update(self.annotations or {})
+        return anno
+
+    def get_data(self):
+        return self.filtered_data.values(
+            *self.get_groupby()).annotate(
+                **self.get_annotations())
+
+    def get_groupby(self):
+        return self.groupby and [self.groupby] or self.default_groupby
+
+
+class TagsDataTool(FilteredData):
+    """Base Data Tool for retrieving Tag data
+    """
+    _default_annotations = ()
+    default_groupby = ('resource__tag', )
+    filter_methods = (
+        'tag', 'locales', 'projects')
+    clone_kwargs = (
+        'locales',
+        'projects',
+        'priority',
+        'slug',
+        'path')
+
+    @property
+    def locale_manager(self):
+        return Locale.objects
+
+    @property
+    def project_manager(self):
+        return Project.objects
+
+    @property
+    def resource_manager(self):
+        return Resource.objects
+
+    @property
+    def tag_manager(self):
+        return Tag.objects
+
+    @property
+    def translation_manager(self):
+        """Django object manager for ``Translation``
+        """
```

Comment:
It's pretty obvious what the code does, but I agree - the comment doesn't hurt. I still wonder why comment here and not for other properties as well?

What would be more interesting to read a comment about is why do we need managers to be available as `TagsDataTool` properites.

Reply:
I think i started adding comments - and thought similar to what you wrote - that the comments werent helping. Im happy to add/remove as you feel.

re overriding the managers - it lets you customize them (theoretically) - im happy to add a comment to that effect - altho i think probs in the class comment or similar